rs/cli/src/idlgen.rs

```diff
@@ -0,0 +1,359 @@
+        drop(fs::copy(from_lock, to_lock));
+
+        let res = cargo_run_bin(&gen_manifest_path, &crate_name, self.target_dir);
+
+        fs::remove_dir_all(crate_dir)?;
+
+        match res {
+            Ok(exit_status) if exit_status.success() => Ok(out_file),
+            Ok(exit_status) => Err(anyhow::anyhow!("Exit status: {}", exit_status)),
+            Err(err) => Err(err),
+        }
+    }
+}
+
+fn is_implement_trait_path(
+    idx: &rustdoc_types::Item,
+    program_meta_id: &rustdoc_types::Id,
+) -> Option<rustdoc_types::Path> {
+    if let rustdoc_types::ItemEnum::Impl(item) = &idx.inner {
+        if let Some(tp) = &item.trait_ {
+            if &tp.id == program_meta_id {
+                if let rustdoc_types::Type::ResolvedPath(path) = &item.for_ {
+                    return Some(path.clone());
+                }
+            }
+        }
+    }
+    None
+}
+
+fn get_crate_name(program_package: &Package) -> String {
+    format!("{}-idl-gen", program_package.name)
+}
+
+fn write_file<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> anyhow::Result<()> {
+    let path = path.as_ref();
+    fs::write(path, contents.as_ref())
+        .with_context(|| format!("failed to write `{}`", path.display()))
+}
+
+fn cargo_doc(
+    manifest_path: &cargo_metadata::camino::Utf8Path,
+    target_dir: &cargo_metadata::camino::Utf8Path,
+) -> anyhow::Result<ExitStatus> {
+    let cargo_path = std::env::var("CARGO").unwrap_or("cargo".into());
+
+    let mut cmd = Command::new(cargo_path);
+    cmd.env("RUSTC_BOOTSTRAP", "1")
+        .env(
+            "RUSTDOCFLAGS",
+            "-Z unstable-options --output-format=json --cap-lints=allow",
+        )
+        .env("__GEAR_WASM_BUILDER_NO_BUILD", "1")
+        .stdout(std::process::Stdio::null()) // Don't pollute output
+        .arg("doc")
+        .arg("--manifest-path")
+        .arg(manifest_path.as_str())
+        .arg("--target-dir")
+        .arg(target_dir.as_str())
+        .arg("--no-deps")
+        .arg("--quiet");
+
+    cmd.status()
+        .context("failed to execute `cargo doc` command")
+}
+
+fn cargo_run_bin(
+    manifest_path: &cargo_metadata::camino::Utf8Path,
+    bin_name: &str,
+    target_dir: &cargo_metadata::camino::Utf8Path,
+) -> anyhow::Result<ExitStatus> {
+    let cargo_path = std::env::var("CARGO").unwrap_or("cargo".into());
+
+    let mut cmd = Command::new(cargo_path);
+    cmd.env("CARGO_TARGET_DIR", target_dir)
+        .env("__GEAR_WASM_BUILDER_NO_BUILD", "1")
+        .stdout(std::process::Stdio::null()) // Don't pollute output
+        .arg("run")
+        .arg("--manifest-path")
+        .arg(manifest_path.as_str())
+        .arg("--bin")
+        .arg(bin_name);
+    cmd.status().context("failed to execute `cargo` command")
+}
+
+enum MetaPathVersion {
+    V1,
+    V2,
+}
+
+pub const META_PATH_V1: &[&str] = &["sails_rs", "meta", "ProgramMeta"];
+pub const META_PATH_V2: &[&str] = &["sails_idl_meta", "ProgramMeta"];
```

Comment:
Do you want this to be implemented like
```
impl MetaPathVersion {
    path(&self) -> &[str] {
        match *self ...
    }
}
```
and the have consts for MetaPathVersion V1 and V2

Reply:
hmm, rafactored